### ISSUE TITLE
tensorflow-1.8.0rc0: tf.compat.as_str returns bytes for python3 since 20180409

### ISSUE THREAD

**Issue by andresusanopinto** [tensorflow/tensorflow#18598]
The issue appeared first in `tf-nightly==1.8.0.dev20180409` but is now present in `tensorflow==1.8.0rc0`.

Reproduce steps:

```
$ python3 -c "import tensorflow as tf; print(tf.VERSION, type(tf.compat.as_str('hello')) == str)
```

Is expected to always print "True".  But gets:

```
# tensorflow
1.6.0 True
1.7.0 True
1.8.0-rc0 False           <= Broken!

# tf-nightly
1.8.0-dev20180408 True
1.8.0-dev20180409 False   <= Broken!
```

**Comment by mrry** [comment]
@annarev When I trace through this code in Python 3, it does indeed look like `tf.compat.as_str()` is mapped to this function:

https://github.com/tensorflow/tensorflow/blob/3128b43eb0bf37ac3c49cb22a6e1789d8ea346e8/tensorflow/python/util/compat.py#L48-L68

However, if I do `from tensorflow.python.util import compat`, `compat.as_str()` maps to `as_text()` via (presumably) this appropriate redirection:

https://github.com/tensorflow/tensorflow/blob/3128b43eb0bf37ac3c49cb22a6e1789d8ea346e8/tensorflow/python/util/compat.py#L93-L97

Is it possible that the `@tf_export()` decorator is rebinding the symbol incorrectly? From what I can tell, it has the correct behavior in Python 2, but should have different behavior in Python 3.

**Comment by yongtang** [comment]
@mrry @annarev I think the tf_export could be explicitly called based on python 2 or 3 to address the issue. The following diff might work:
```diff
diff --git a/tensorflow/python/util/compat.py b/tensorflow/python/util/compat.py
index 4163fca..738479c 100644
--- a/tensorflow/python/util/compat.py
+++ b/tensorflow/python/util/compat.py
@@ -45,7 +45,6 @@ from tensorflow.python.util.tf_export import tf_export
 from tensorflow.python.util.tf_export import tf_export
 
 
-@tf_export('compat.as_bytes', 'compat.as_str')
 def as_bytes(bytes_or_text, encoding='utf-8'):
   """Converts either bytes or unicode to `bytes`, using utf-8 encoding for text.
 
@@ -68,7 +67,6 @@ def as_bytes(bytes_or_text, encoding='utf-8'):
                     (bytes_or_text,))
 
 
-@tf_export('compat.as_text')
 def as_text(bytes_or_text, encoding='utf-8'):
   """Returns the given argument as a unicode string.
 
@@ -93,8 +91,12 @@ def as_text(bytes_or_text, encoding='utf-8'):
 # Convert an object to a `str` in both Python 2 and 3.
 if _six.PY2:
   as_str = as_bytes
+  tf_export('compat.as_bytes', 'compat.as_str')(as_bytes)
+  tf_export('compat.as_text')(as_text)
 else:
   as_str = as_text
+  tf_export('compat.as_bytes')(as_bytes)
+  tf_export('compat.as_text', 'compat.as_str')(as_text)
 
 
 @tf_export('compat.as_str_any')
```

**Comment by andresusanopinto** [comment]
Can we get this merged in 1.8? 

https://github.com/tensorflow/tensorflow/blob/r1.8/tensorflow/python/util/compat.py

**Comment by andresusanopinto** [comment]
Thanks @yongtang!
